chain/chain/src/resharding/manager.rs

```diff
@@ -265,45 +229,47 @@ impl ReshardingManager {
                 retain_mode,
             )?;
 
+            // Split the parent trie and create a new child trie.
+            let trie_changes =
+                parent_trie.retain_split_shard(&split_shard_event.boundary_account, retain_mode)?;
+            let new_root = tries.apply_all(&trie_changes, parent_shard_uid, &mut store_update);
+            tries.apply_memtrie_changes(&trie_changes, parent_shard_uid, block_height);
+
+            // TODO(resharding): remove duplicate_nodes_at_split_boundary method after proper fix for refcount issue
             let trie_recorder = parent_trie.take_recorder().unwrap();
-            let partial_storage = trie_recorder.write().expect("no poison").recorded_storage();
-            let partial_state_len = match &partial_storage.nodes {
-                PartialState::TrieValues(values) => values.len(),
-            };
+            let mut trie_recorder = trie_recorder.write().expect("no poison");
+            Self::duplicate_nodes_at_split_boundary(
+                &mut store_update,
+                trie_recorder.recorded_iter(),
+                parent_shard_uid,
+            );
 
             // TODO(resharding): set all fields of `ChunkExtra`. Consider stronger
             // typing. Clarify where it should happen when `State` and
             // `FlatState` update is implemented.
             let mut child_chunk_extra = ChunkExtra::clone(&parent_chunk_extra);
-            *child_chunk_extra.state_root_mut() = new_state_root;
+            *child_chunk_extra.state_root_mut() = new_root;
             *child_chunk_extra.congestion_info_mut() = child_congestion_info;
 
             chain_store_update.save_chunk_extra(block_hash, &new_shard_uid, child_chunk_extra);
             chain_store_update.save_state_transition_data(
                 *block_hash,
                 new_shard_uid.shard_id(),
-                Some(partial_storage),
+                Some(trie_recorder.recorded_storage()),
                 CryptoHash::default(),
                 // No contract code is accessed or deployed during resharding.
                 // TODO(#11099): Confirm if sending no contracts is ok here.
                 Default::default(),
             );
 
-            // Commit `TrieChanges` directly. They are needed to serve reads of
-            // new nodes from `DBCol::State` while memtrie is properly created
-            // from flat storage.
-            tries.apply_insertions(
-                &trie_changes,
-                new_shard_uid,
-                &mut trie_store_update.trie_store_update(),
-            );
-            tracing::info!(
-                target: "resharding", ?new_shard_uid, ?new_state_root, ?partial_state_len,
-                "Child memtrie created"
-            );
+            tracing::info!(target: "resharding", ?new_shard_uid, ?new_root, "Splitting parent memtrie done");
```

Comment:
Since there will be two such logs (one for each child) I like the current message a bit better.